compiler/src/iree/compiler/Dialect/Stream/Transforms/AutomaticReferenceCounting.cpp

```diff
@@ -0,0 +1,728 @@
+        llvm::dbgs() << " failed: predecessor not local\n";
+      });
+      return false;
+    } else if (timepointIt == timepoints.end()) {
+      LLVM_DEBUG({
+        llvm::dbgs() << "[arc] coverage check for predecessor ";
+        predecessor.printAsOperand(llvm::dbgs(), *asmState);
+        llvm::dbgs() << " over ";
+        timepoint.printAsOperand(llvm::dbgs(), *asmState);
+        llvm::dbgs() << " failed: timepoint not local\n";
+      });
+      return false;
+    }
+    return map[timepointIt->second].test(predecessorIt->second);
+  }
+
+  // Returns the unique assigned index for |timepoint|, assigning one if needed.
+  unsigned getOrInsertIndex(Value timepoint) {
+    auto it = timepoints.find(timepoint);
+    if (it != timepoints.end()) {
+      return it->second;
+    }
+    unsigned index = timepoints.size();
+    timepoints.insert({timepoint, index});
+    map.resize(index + 1);
+    auto &set = map[index];
+    set.resize(index + 1);
+    set.set(index); // identity
+    return index;
+  }
+};
+
+using TimepointSet = SmallPtrSet<Value, 4>;
+
+// Manages a set of per-resource timeline uses.
+// These are not the same as the last users of any particular SSA value as
+// resources may be aliased and the timeline may be constructed out of order.
+// In cases of forks in the timeline there may be multiple last uses.
+struct LastUseSet {
+  // Must be provided if LLVM_DEBUG is enabled.
+  AsmState *asmState;
+  // Timepoint coverage map.
+  LocalTimepointCoverage &coverage;
+  // Resource base value to a set of signal timepoints from users.
+  // Each set of timepoints is maintained as only those not covered by
+  // others. Multiple values indicates a fork.
+  DenseMap<Value, TimepointSet> map;
+  // Map of resource values back to the root resource value they alias.
+  // Each new alias added is always to the first root identified.
+  DenseMap<Value, Value> aliases;
+  // All base resources in the order they occur in the block.
+  SmallVector<Value> baseResourceOrder;
+
+  LastUseSet() = delete;
+  LastUseSet(AsmState *asmState, LocalTimepointCoverage &coverage)
+      : asmState(asmState), coverage(coverage) {}
+
+  // Calls |fn| for each base resource produced within the analysis scope with
+  // the set of timepoints indicating when the last uses have completed. After
+  // all timepoints in the set have been reached the resource may be
+  // deallocated.
+  void forEachResource(
+      std::function<void(Value resource, TimepointSet &timepoints)> fn) {
+    for (auto baseResource : llvm::reverse(baseResourceOrder)) {
+      fn(baseResource, map[baseResource]);
+    }
+  }
+
+  // Inserts a produced |resource| into the timepoint set ready when
+  // |fromTimepoint| is reached. A new entry will be created for the
+  // resource in the map.
+  void produce(Value resource, Value fromTimepoint) {
+    Value baseResource = lookupResource(resource);
+    assert(!map.contains(baseResource) && "produced values must be unique");
```

Comment:
Could anything marked `produce` have a base resource that it aliases?

Reply:
nope! I was too defensive here - I've removed the lookup and made it an assertion instead